### ISSUE TITLE
Latency setting ?

### ISSUE THREAD

**Issue by Hierosme** [hrsh7th/nvim-cmp#715]
### FAQ

- [X] I have checked the [FAQ](https://github.com/hrsh7th/nvim-cmp#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/hrsh7th/nvim-cmp/issues) and there are no open or closed issues with the same problem.

### Neovim Version

NVIM v0.6.0 Build type: Release LuaJIT 2.1.0-beta3 Compilé par runner@fv-az32-74  Features: +acl +iconv +tui See ":help feature-compile"           fichier vimrc système : "$VIM/sysinit.vim"                $VIM par défaut : "/share/nvim"  Run :checkhealth for more info

### Minimal reproducible config

Go to the on line documentation , use you Browser search command. And seach for **Latency**

### Description

The completion pop qnd overlap evrything without any explaination on the documentation.

Please, reallly Please for you users. Clarify how deal with completion latency. the  original updatetime is si;ply ignored by nvim-cmp

### Steps to reproduce

All ready explain

### Expected behavior

Not related

### Actual behavior

Not related

### Additional context

nvim-cmp work well but i have totaly no control, because updatetime is ignored

**Comment by Shougo** [comment]
It seems feature request than bug.
It means completion delay feature for auto completion?

**Comment by Hierosme** [comment]
yes , or if all ready exist just a documentation section

Thanks for you assistance @Shougo

**Comment by Shougo** [comment]
I have read the documentation and the feature is not implemented in nvim-cmp.

**Comment by hrsh7th** [comment]
I have received this request several times so far.
But I don't implement this because I don't know why this setting is useful.

**Comment by Hubro** [comment]
I was looking for this setting in relation to cmp-cmdline. Whenever I very quickly write `:w<CR>` or `:wq<CR>`, the completion menu pops up for a millisecond. This is kind of annoying. I wish I could set a ~300ms delay before the completion window popped up, just for the cmdline completions.

Currently I'm working around this by setting `completion.keyword_length = 3`, but it's not the same. If I type `:ta` and wait, I want suggestions for all the tab related commands I could write. I can just summon the completion window myself by pressing `<Tab>`, of course. This is all just slightly less ideal than being able to set a small delay :smile: 

I'm sure there are also people who just want to type without the completion menu popping up, unless they pause for a second or two. I don't really care about that though.

**Comment by hierosme-nova** [comment]
i confirme,

Sometime the completion distract me, i just wait have a slow completion.

It existe a other case where i just put my cursor somewhere and the completion hide what i originaly want to read.

From my point if i want a speed completion i spam my TAB key..

Regards

**Comment by hrsh7th** [comment]
```
let s:timer = 0
autocmd TextChangedI * call s:on_text_changed()
function! s:on_text_changed() abort
  call timer_stop(s:timer)
  let s:timer = timer_start(500, function('s:complete'))
endfunction
function! s:complete() abort
  lua require('cmp').complete({ reason = require('cmp').ContextReacton.Auto })
endfunction
```

**Comment by joeyschoblaska** [comment]
@hrsh7th Is there a way to prevent this from firing when buftype is "prompt"?

**Comment by towry** [comment]
> @hrsh7th Is there a way to prevent this from firing when buftype is "prompt"?

see https://github.com/hrsh7th/nvim-cmp/issues/60#issuecomment-1247666586

**Comment by hrsh7th** [comment]
You can check `&buftype ~= 'prompt'`

**Comment by towry** [comment]
I turned it to lua and it is working for delay completion, but it seems also **disabled the cmd completion**, and the `enabled` option is not working too(used to filter out 'prompt' buftype)

```lua
 -- local timer = 0
  -- local function do_complete()
  --   cmp.complete({ reason = cmp.ContextReason.Auto })
  -- end
  -- vim.api.nvim_create_autocmd(
  --   { "TextChangedI", },
  --   {
  --     callback = function()
  --       vim.fn.timer_stop(timer)
  --       timer = vim.fn.timer_start(500, do_complete)
  --     end,
  --     pattern = "*",
  --   }
  -- )
```

**Comment by Hierosme** [comment]
Hello,

i continue to follow it feature request ...

I have to repport a issue of the fire completion with the shadow text experimental option.

If shadow text is enable the fire completion is a crime because a simple "\n" select the first item. Unfortunally i just want go to the next line.

Exemple:
def foo():     <--------------------------------------- Here

I just want go to the next line, but the fire completion select me the frist item and write everything execpt what i want.
Actually i fight hard with the fire completion each time i write a new function. (Esc, mode insert, Enter)


I understand the rule about too many setting is the evil, then if the better is to limit the user setting requirement, my point is to fixe the fire completion trouble  with a  internal priority of the completion source + response delay of that source.

That is clear time to time my LSP is slow and the completion dont care about LSP  delay responses.
Other time the LSP is ready and the completion dont make any mess, by using snippet or buffer source ...

Regards

**Comment by pinggit** [comment]
it looks like this is still not implemented.
I think lacking of latency setting is the blocker for me to use this plugin. the suggestion comes too fast and always distractive, have to remove it :(

**Comment by Hubro** [comment]
@pinggit Scroll up to [hrsh7th](https://github.com/hrsh7th)'s response on March 16th last year, it contains an example snippet for how to delay the completion popup.

**Comment by pinggit** [comment]
@Hubro thanks.
I copy and paste the mentioned config in my .vimrc, and I got this error:
```
[string "vim chunk"]:1: module 'cmp' not found:
```

**Comment by scienceplease** [comment]
For anyone stumbling across the need for this and has a Lua-based config, this is the gist of how I achieved it -- details are omitted for brevity. It's not enough to just insert a delay, you need to track whether the cursor has moved before you actually open the completion menu.

```lua
-- Somewhere before you setup nvim-cmp.
_G.user.cmp = {
    cursor_tick = { }, -- Per-buffer counter to track whether cursor has moved.
}

...

-- Autocommands specific to nvim-cmp.
function M.autocmds()

    local augroup = vim.api.nvim_create_augroup
    local autocmd = vim.api.nvim_create_autocmd

    local UserCmp = augroup('UserCmp', { clear = true })

    local state   = _G.user.cmp
    local completion_menu_delay = 600  -- ms

    autocmd('TextChangedI', {
        desc  = 'Delay opening nvim-cmp completion menu in normal buffers.',
        group = UserCmp,
        callback = function(opts)
            if vim.bo.buftype == '' then
                local tick_saved = state.cursor_tick[opts.buf]

                vim.defer_fn(
                    function()
                        if state.cursor_tick[opts.buf] == tick_saved then
                            -- Only open completion menu if cursor has not moved since.
                            cmp.complete({ reason = cmp.ContextReason.Auto })
                        end
                    end,
                    completion_menu_delay)
            else
                -- No delay for non-normal buffer types.
                cmp.complete({ reason = cmp.ContextReason.Auto })
            end
        end,
    })

    autocmd('CursorMovedI', {
        desc  = 'Increment buffer cursor tick on cursor move.',
        group = UserCmp,
        callback = function(opts)
            if vim.bo.buftype == '' then
                state.cursor_tick[opts.buf] = state.cursor_tick[opts.buf] + 1
            end
        end,
    })

    autocmd('BufEnter', {
        desc  = 'Reset cursor tick for buffer on enter.',
        group = UserCmp,
        callback = function(opts)
            if vim.bo.buftype == '' then
                state.cursor_tick[opts.buf] = 0
            end
        end,
    })
end

...

function M.buf_autocomplete()
-- Called on per-buffer basis to determine whether completion menu is automatically opened.

    if vim.fn.mode() == 'c' then
        return true   -- Always enable in COMMAND mode.
    elseif vim.bo.buftype == '' then
        return false  -- Explicitly open completion menu on 'TextChangedI' for normal buffers via autocommand (to insert a delay).
    elseif vim.bo.buftype == 'prompt' and vim.bo.filetype == 'TelescopePrompt' then
        return false  -- Permanently disable for telescope prompt.
    else
        return true
    end
end

...

-- nvim-cmp settings table
{
  enabled = M.buf_autocomplete,
...
}
```

**Comment by WaitingForGuacamole** [comment]
I like @kekscode solution, and did some minor tweaks to give you control over the delay (either from other Lua code or the command line:

```
local completionDelay = 750
local timer = nil
local cmp = require("cmp")

function _G.setAutoCompleteDelay(delay)
  completionDelay = delay
end

function _G.getAutoCompleteDelay()
  return completionDelay
end

vim.api.nvim_create_autocmd({ "TextChangedI", "CmdlineChanged" }, {
  pattern = "*",
  callback = function()
    if timer then
      vim.loop.timer_stop(timer)
      timer = nil
    end

    timer = vim.loop.new_timer()
    timer:start(
      _G.getAutoCompleteDelay(),
      0,
      vim.schedule_wrap(function()
        cmp.complete({ reason = cmp.ContextReason.Auto })
      end)
    )
  end,
})
```

So you can change the delay with:

`:lua setAutoCompleteDelay(300)`

or get the current setting with 

`:lua print(getAutoCompleteDelay())`

**Comment by hamishimac** [comment]
The auto-popup is annoying when it pops up *right* before you type enter (or whatever key accepts the completion). This is the situation where a built-in auto-complete latency (option) would be helpful.  

I investigated the above examples, but they seem kind of hacky, since they don't seem to be aware of the state of completion.  An illustration of this is the following.  Let's say I select an autocompletion after it pops up.  After the autocompletion is inserted, the autocompletion pops up again.  If I haven't typed anything after I accepted a previous autocompletion, I wouldn't expect the autocompletion window to pop up.
